cloud-controller-manager/do/firewall_controller_test.go

```diff
@@ -0,0 +1,420 @@
+			fwCache:              newFakeFirewallCache(fakeWorkerFWName, fakeInboundRule),
+			expectedGetResponse:  newFakeNotFoundResponse(),
+			expectedListResponse: newFakeOKResponse(),
+			expectedListResult:   []godo.Firewall{*newFakeFirewall(fakeWorkerFWName, fakeInboundRule)},
+			expectedFirewall:     newFakeFirewall(fakeWorkerFWName, fakeInboundRule),
+		},
+		{
+			name:                "get firewall from API with cached firewall ID",
+			fwCache:             newFakeFirewallCache(fakeWorkerFWName, fakeInboundRule),
+			expectedGetResponse: newFakeOKResponse(),
+			expectedGetResult:   newFakeFirewall(fakeWorkerFWName, fakeInboundRule),
+			expectedFirewall:    newFakeFirewall(fakeWorkerFWName, fakeInboundRule),
+		},
+		{
+			name:                 "return firewall from list when it matches the firewall name",
+			expectedFirewall:     newFakeFirewall(fakeWorkerFWName, fakeInboundRule),
+			fwCache:              newFakeFirewallCache(fakeWorkerFWName, fakeInboundRule),
+			expectedGetResponse:  newFakeNotFoundResponse(),
+			expectedListResponse: newFakeOKResponse(),
+			expectedListResult:   []godo.Firewall{*newFakeFirewall(fakeWorkerFWName, fakeInboundRule)},
+		},
+	}
+	for _, test := range testcases {
+		t.Run(test.name, func(t *testing.T) {
+			gclient := newFakeGodoClient(
+				&fakeFirewallService{
+					getFunc: func(context.Context, string) (*godo.Firewall, *godo.Response, error) {
+						return test.expectedGetResult, test.expectedGetResponse, test.expectedGetError
+					},
+					listFunc: func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error) {
+						return test.expectedListResult, test.expectedListResponse, test.expectedListError
+					},
+				},
+			)
+			fwManagerOp = newFakeFirewallManagerOp(gclient, test.fwCache)
+
+			fw, err := fwManagerOp.Get(ctx)
+			if want, got := test.expectedError, err; !reflect.DeepEqual(want, got) {
+				t.Errorf("incorrect firewall config\nwant: %#v\n got: %#v", want, got)
+			}
+
+			if !cmp.Equal(fw, test.expectedFirewall) {
+				t.Errorf("got %q, want %q", fw, test.expectedFirewall)
+			}
+		})
+	}
+}
+
+func TestFirewallController_Set(t *testing.T) {
+	testcases := []struct {
+		name         string
+		fwCache      firewallCache
+		inboundRules []godo.InboundRule
+		firewall     *godo.Firewall
+		// Expected DO Firewall API, AddRules request, response values.
+		expectedAddRulesResponse *godo.Response
+		expectedAddRulesError    error
+		// Expected DO Firewall API, Create request, response values.
+		expectedCreateResult   *godo.Firewall
+		expectedCreateResponse *godo.Response
+		expectedCreateError    error
+		// Expected DO Firewall API, Get request, response values.
+		expectedGetResult   *godo.Firewall
+		expectedGetResponse *godo.Response
+		expectedGetError    error
+		expectedError       error
+		// Expected DO Firewall API, List request, response values.
+		expectedListResult   *godo.Firewall
```

Comment:
I'd match this with the type we use for testing `Get()`, which is a `[]godo.Firewall`, allowing us to stub out an empty list properly.